lib/Dialect/TritonGPU/Transforms/HoistTMEMAlloc.cpp

```diff
@@ -92,6 +92,56 @@ class RemoveUnusedTMEMLoad : public OpRewritePattern<TMEMTokenLoadOp> {
   }
 };
 
+std::optional<bool> getBoolFromConstant(Value cst) {
+  auto constantOp = cst.getDefiningOp<arith::ConstantOp>();
+  if (!constantOp) {
+    return std::nullopt;
+  }
+  assert(constantOp.getValue());
+  if (auto boolAttr = dyn_cast<BoolAttr>(constantOp.getValue())) {
+    return boolAttr.getValue();
+  }
+  return std::nullopt;
+}
+
+class RemoveUnusedTMEMStore : public OpRewritePattern<TMEMTokenStoreOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(TMEMTokenStoreOp store,
+                                PatternRewriter &rewriter) const override {
+    auto tok = store.getToken();
+    if (!tok.hasOneUse())
+      return failure();
+    auto loop = dyn_cast<scf::ForOp>(*tok.getUsers().begin());
+    if (!loop)
+      return failure();
+    auto loopTok = loop.getBody()->getArgument(
+        tok.getUses().begin()->getOperandNumber() - 2);
+    if (!loopTok.hasOneUse())
+      return failure();
+    auto mma =
+        dyn_cast<nvidia_gpu::MMAv5OpInterface>(*loopTok.getUsers().begin());
+    if (!mma)
+      return failure();
+    auto useD = dyn_cast<BlockArgument>(mma.useAccumulator());
+    if (!useD)
+      return failure();
+    auto parent = useD.getParentBlock()->getParentOp();
+    if (parent != loop)
+      return failure();
+    auto loopInit = loop.getInitArgs()[useD.getArgNumber() - 1];
+    auto val = getBoolFromConstant(loopInit);
+    if (!val)
+      return failure();
+    if (val.value() == true)
+      return failure();
+    rewriter.replaceAllUsesWith(store.getToken(), store.getDep());
```

Comment:
is that correct in case the loop has zero iterations? in this case the mma with init acc would never be reached?

Reply:
mmm.

Yes, it is technically possible a user could pass in something that causes a dynamically calculated loop boundary to fail to execute, in which case, the tmem would never be initialized.

Reply:
then we would pick up undefined values outside the loop instead of 0. That's the reason why we haven't been removing this kind of cases. I'm not sure we have a way to properly doing this without making assumptions on the loop bounds.

Reply:
I just pushed a change to add an conditional after the loop that stores if the runtime loop boundaries mean that no iterations happen.

PTAL.

Reply:
sounds good, looks good once the other comments are addressed